[BUG] APIM policies rules parsing policies with C# code  fail to convert deployable XML into type System.XML.XmlDocument

### Existing rule

Azure.APIM.CORSPolicy and Azure.APIM.PolicyBase

### Description of the issue

When running PSRule against APIM policies that sucessfully deploys on Azure, the rule fails due to being unable to convert the XML of the policy

### Error messages

```text

##[error]AZR-000371: apim-name/api-name failed Azure.APIM.PolicyBase with the error 'Cannot convert value "<!-- Add policies as children to the <inbound>, <outbound>, <backend>, and <on-error> elements -->
<policies>
	<!-- Throttle, authorize, validate, cache, or transform the requests -->
	<inbound>
		<base />
		<set-backend-service base-url="{{backend}}" />
		<authentication-managed-identity resource="{{audiencee}}" />
		<set-variable name="originalHost" value="@(context.Rquest.Headers.GetValueOrDefault("X-Original-Host", "NotAvailable"))" />
	</inbound>
	<!-- Control if and how the requests are forwarded to services  -->
	<backend>
		<base />
	</backend>
	<!-- Customize the responses -->
	<outbound>
		<base />
	</outbound>
	<!-- Handle exceptions and customize error responses  -->
	<on-error>
		<base />
	</on-error>
</policies>" to type "System.Xml.XmlDocument". Error: "'X-Original-Host' is an unexpected token. Expecting whitespace. Line 15, position 89."'.

At /usr/local/share/powershell/Modules/PSRule.Rules.Azure/1.39.3/rules/Azure.APIM.Rule.ps1:384 char:21
+         $policies | ForEach-Object {
+                     ~~~~~~~~~~~~~~~~

at <ScriptBlock>, /usr/local/share/powershell/Modules/PSRule.Rules.Azure/1.39.3/rules/Azure.APIM.Rule.ps1: line 386
at global:GetAPIMPolicyNode<Process>, /usr/local/share/powershell/Modules/PSRule.Rules.Azure/1.39.3/rules/Azure.APIM.Rule.ps1: line 384
at <ScriptBlock>, /usr/local/share/powershell/Modules/PSRule.Rules.Azure/1.39.3/rules/Azure.APIM.Rule.ps1: line 326
at Rule 'Azure.APIM.PolicyBase', /usr/local/share/powershell/Modules/PSRule.Rules.Azure/1.39.3/rules/Azure.APIM.Rule.ps1: line 325


```

### Reproduction

Try to run PSRule against a bicep template that deploys the following policy for an API with `rawxml` as the format

```
<!--
    - Policies are applied in the order they appear.
    - Position <base/> inside a section to inherit policies from the outer scope.
    - Comments within policies are not preserved.
-->
<!-- Add policies as children to the <inbound>, <outbound>, <backend>, and <on-error> elements -->
<policies>
	<!-- Throttle, authorize, validate, cache, or transform the requests -->
	<inbound>
		<base />
		<set-backend-service base-url="{{backend}}" />
		<authentication-managed-identity resource="{{audiencee}}" />
		<set-variable name="originalHost" value="@(context.Rquest.Headers.GetValueOrDefault("X-Original-Host", "NotAvailable"))" />
	</inbound>
	<!-- Control if and how the requests are forwarded to services  -->
	<backend>
		<base />
	</backend>
	<!-- Customize the responses -->
	<outbound>
		<base />
	</outbound>
	<!-- Handle exceptions and customize error responses  -->
	<on-error>
		<base />
	</on-error>
</policies>
```

The rule will fails but a deployment of the template to Azure will succeed

### Version of PSRule

2.9.0

### Version of PSRule for Azure

1.39.3

### Additional context

_No response_

--- a/docs/CHANGELOG-v1.md
+++ b/docs/CHANGELOG-v1.md
@@ -29,6 +29,12 @@ See [upgrade notes][1] for helpful information when upgrading from previous vers
 
 ## Unreleased
 
+What's changed since pre-release v1.40.0-B0147:
+
+- Bug fixes:
+  - Fixed evaluation of APIM policies when using embedded C# with quotes by #BernieWhite.
+    [#3184](https://github.com/Azure/PSRule.Rules.Azure/issues/3184)
+
 ## v1.40.0-B0147 (pre-release)
 
 What's changed since pre-release v1.40.0-B0103:

--- a/src/PSRule.Rules.Azure/Data/APIM/APIMPolicyReader.cs
+++ b/src/PSRule.Rules.Azure/Data/APIM/APIMPolicyReader.cs
@@ -1,0 +1,34 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System.Xml;
+using System.Text.RegularExpressions;
+using System.IO;
+
+namespace PSRule.Rules.Azure.Data.APIM;
+
+/// <summary>
+/// A reader for APIM policy files.
+/// </summary>
+internal static class APIMPolicyReader
+{
+    // Create a regex pattern to match the `="@(` `)"` pairs.
+    private static readonly Regex _Pattern = new(@"(?<=\=""\@\().*?(?=\)"")", RegexOptions.Singleline | RegexOptions.Compiled);
+
+    /// <summary>
+    /// Read the content of the policy file as XML.
+    /// </summary>
+    /// <remarks>
+    /// This method automatically escapes quotes within policy expressions to ensure the XML is well-formed.
+    /// </remarks>
+    public static XmlReader ReadContent(string content)
+    {
+        // For each match replace `"` characters with `&quot;`.
+        foreach (Match match in _Pattern.Matches(content))
+        {
+            content = content.Replace(match.Value, match.Value.Replace("\"", "&quot;"));
+        }
+
+        return XmlReader.Create(new StringReader(content));
+    }
+}

--- a/src/PSRule.Rules.Azure/Runtime/Helper.cs
+++ b/src/PSRule.Rules.Azure/Runtime/Helper.cs
@@ -5,8 +5,10 @@ using System;
 using System.IO;
 using System.Linq;
 using System.Management.Automation;
+using System.Xml;
 using PSRule.Rules.Azure.Configuration;
 using PSRule.Rules.Azure.Data;
+using PSRule.Rules.Azure.Data.APIM;
 using PSRule.Rules.Azure.Data.Bicep;
 using PSRule.Rules.Azure.Data.Network;
 using PSRule.Rules.Azure.Data.Template;
@@ -204,6 +206,17 @@ public static class Helper
 
         var parts = resourceName.Split('/');
         return parts[parts.Length - 1];
+    }
+
+    /// <summary>
+    /// Load a APIM policy as an XML document.
+    /// </summary>
+    public static XmlDocument GetAPIMPolicyDocument(string content)
+    {
+        using var reader = APIMPolicyReader.ReadContent(content);
+        var doc = new XmlDocument();
+        doc.Load(reader);
+        return doc;
     }
 
     #region Helper methods

--- a/src/PSRule.Rules.Azure/rules/Azure.APIM.Rule.ps1
+++ b/src/PSRule.Rules.Azure/rules/Azure.APIM.Rule.ps1
@@ -383,7 +383,7 @@ function global:GetAPIMPolicyNode {
         }
         $policies | ForEach-Object {
             if (!($IgnoreGlobal -and $_.type -eq 'Microsoft.ApiManagement/service/policies') -and $_.properties.format -in 'rawxml', 'xml' -and $_.properties.value) {
-                $xml = [Xml]$_.properties.value
+                $xml = [PSRule.Rules.Azure.Runtime.Helper]::GetAPIMPolicyDocument($_.properties.value)
                 $xml.SelectNodes("//${Node}")
             }
         }

--- a/tests/PSRule.Rules.Azure.Tests/APIM/APIMPolicyTests.cs
+++ b/tests/PSRule.Rules.Azure.Tests/APIM/APIMPolicyTests.cs
@@ -1,0 +1,39 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System.Xml;
+using PSRule.Rules.Azure.Data.APIM;
+
+namespace PSRule.Rules.Azure.APIM;
+
+/// <summary>
+/// Test cases for APIM policy files with complex syntax.
+/// </summary>
+public sealed class APIMPolicyTests : BaseTests
+{
+    /// <summary>
+    /// Test case for https://github.com/Azure/PSRule.Rules.Azure/issues/3184
+    /// </summary>
+    [Fact]
+    public void APIMPolicyReader_WhenExpressionIsFound_ShouldEscapeAndLoad()
+    {
+        using var reader = ReadContentFromFile("APIM/Tests.Policy.1.xml");
+
+        var doc = new XmlDocument();
+        doc.Load(reader);
+
+        var node = doc.SelectSingleNode("//set-variable");
+
+        Assert.Equal("@(context.Request.Headers.GetValueOrDefault(\"X-Original-Host\", \"NotAvailable\"))", node.Attributes["value"].Value);
+    }
+
+    #region Helper methods
+
+    private static XmlReader ReadContentFromFile(string fileName)
+    {
+        var content = GetContent(fileName);
+        return APIMPolicyReader.ReadContent(content);
+    }
+
+    #endregion Helper methods
+}

--- a/tests/PSRule.Rules.Azure.Tests/APIM/Tests.Policy.1.xml
+++ b/tests/PSRule.Rules.Azure.Tests/APIM/Tests.Policy.1.xml
@@ -1,0 +1,19 @@
+<!-- Test case for https://github.com/Azure/PSRule.Rules.Azure/issues/3184 -->
+<!-- Based on work contributed by @GABRIELNGBTUC -->
+<policies>
+	<inbound>
+		<base />
+		<set-backend-service base-url="{{backend}}" />
+		<authentication-managed-identity resource="{{audience}}" />
+    <set-variable name="originalHost" value="@(context.Request.Headers.GetValueOrDefault("X-Original-Host", "NotAvailable"))" />
+	</inbound>
+	<backend>
+		<base />
+	</backend>
+	<outbound>
+		<base />
+	</outbound>
+	<on-error>
+		<base />
+	</on-error>
+</policies>

--- a/tests/PSRule.Rules.Azure.Tests/BaseTests.cs
+++ b/tests/PSRule.Rules.Azure.Tests/BaseTests.cs
@@ -1,0 +1,21 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.IO;
+
+namespace PSRule.Rules.Azure;
+
+public abstract class BaseTests
+{
+    protected static string GetSourcePath(string fileName)
+    {
+        return Path.Combine(AppDomain.CurrentDomain.BaseDirectory, fileName);
+    }
+
+    protected static string GetContent(string fileName)
+    {
+        var path = GetSourcePath(fileName);
+        return File.ReadAllText(path);
+    }
+}

--- a/tests/PSRule.Rules.Azure.Tests/PSRule.Rules.Azure.Tests.csproj
+++ b/tests/PSRule.Rules.Azure.Tests/PSRule.Rules.Azure.Tests.csproj
@@ -317,6 +317,9 @@
     <None Update="Bicep\ScopeTestCases\Tests.Bicep.1.json">
       <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
     </None>
+    <None Update="APIM\Tests.Policy.1.xml">
+      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
+    </None>
   </ItemGroup>
 
 </Project>

--- a/tests/PSRule.Rules.Azure.Tests/TemplateVisitorTestsBase.cs
+++ b/tests/PSRule.Rules.Azure.Tests/TemplateVisitorTestsBase.cs
@@ -1,8 +1,6 @@
 // Copyright (c) Microsoft Corporation.
 // Licensed under the MIT License.
 
-using System;
-using System.IO;
 using System.Linq;
 using Newtonsoft.Json.Linq;
 using PSRule.Rules.Azure.Configuration;
@@ -12,14 +10,9 @@ using static PSRule.Rules.Azure.Data.Template.TemplateVisitor;
 
 namespace PSRule.Rules.Azure;
 
-public abstract class TemplateVisitorTestsBase
+public abstract class TemplateVisitorTestsBase : BaseTests
 {
     #region Helper methods
-
-    protected static string GetSourcePath(string fileName)
-    {
-        return Path.Combine(AppDomain.CurrentDomain.BaseDirectory, fileName);
-    }
 
     protected static JObject[] ProcessTemplate(string templateFile, string parametersFile)
     {

